### DISCUSSION TITLE
Matlab crashing during max(calcDensity(...

### DISCUSSION THREAD

**Discussion by zmichels** [mtex-toolbox/mtex#1132] [Ask Anything]
Hi everybody,

I have been using the following syntax lately to compute a dominant direction from a set of vectors, where the variable 'vecs' is the set of vectors:
`[~,V] = max(calcDensity(vecs))`

Specifically, I am using a set of vectors that are the result of the longAxis function applied to a subset of indexed grains ('g') from my dataset, as so:
`vecs = g.longAxis`

For some subsets (separated using a grain diameter threshold), Matlab crashes when the max() function is applied to the S2FunHarmonic that results from the calcDensity() function. For some subsets, it works fine and Matlab does not crash.

However, if I do the following and use a plotS2Grid ('r'), Matlab never crashes:
```
r = plotS2Grid('resolution',1*degree,'antipodal');
kde = calcDensity([g.longAxis],r,'halfwidth',10*degree);
[~,I]=max(kde)
V = r(I)
```

I would prefer to use the syntax with fewer lines, and sometimes I also want to use the S2FunHarmoinc for other analyses. However, I do not prefer to have Matlab crash unpredictably. I can confirm that this same thing is happening on a colleague's computer when they process the same datasets with the same syntax. So, I don't think it is something unique to my computer/setup.

I am running MTEX 5.7.0 on a Mac with Matlab R2020a. 

Can anyone think of a reason this is happening and/or why the behavior may be so unpredictable-yet-repeatable (i.e., I can't predict beforehand which set of vectors will cause a crash, but any set that does cause a crash will cause a crash every time)?

Thanks for any suggestions.

Cheers,
Zach

**Comment by kilir** [comment]
Hi Zach,
once you have such a set of vectors which crash, could you save it as a .mat and provide for testing?
Cheers,
Rüdiger

**Reply by zmichels** [reply]
This link below is to a set of vectors that leads to a crash. I can successfully compute an S2FunHarmonic, but the crash happens when using max() on that.

https://drive.google.com/file/d/1Fi3WbPTHLZHoNwE9LVpHYCC8ZUXDbZ-o/view?usp=sharing

**Reply by kilir** [reply]
Hi Zach,
no crash for me doing
```
max(S2Crash)
% ans =
%     7.2525

%or
s2f = vCrash.calcDensity
s2f.max
% ans =
%     7.2525
```
neither on Matlab 2021a or 2020b, Mtex latest from git.

To narrow down the problem, does your colleague also use a Mac?

Do you use the pre-compiled mex binaries of nfft or did you build those yourself?

Cheers,
Rüdiger

**Reply by ralfHielscher** [reply]
Hi Zach,

I have looked at your data as well with no problems at all. What is quite strange is that 
you are able to compute the S2FunHarmonic which already requires a call to the NFFT. 
Two questions:
 1. Does ```plot(S2Crash)``` work for you?
 2. Can you try ```S2Crash.bandwidth=32``` and then the max command?

Ralf.

**Comment by zmichels** [comment]
Thanks. Yes, I will do that.

Cheers,
Zach

**Comment by lisakemp** [comment]
Zachary and MTEX community –

Lisa Kempler from MathWorks here.
I’ve been following this discussion.

Zachary – Regardless of what vectors you’re supplying as inputs, MATLAB shouldn’t crash. Error out, sure.  MathWorks would be interested to learn about your particular issues.

Once you have saved the .mat files, as Rudiger requested, it would be ideal if you could submit a bug report on this to MathWorks, along with the relevant files.

Here’s a good place to report a bug: https://www.mathworks.com/support/contact_us.html

Let me know if you have other questions.

Thanks,
Lisa

**Reply by zmichels** [reply]
Hi Lisa,
I submitted a bug report at that link just now.
Thanks,
Zach

**Reply by zmichels** [reply]
Hi Lisa, Ralf, and Rüdiger,

Just an update:

I received this email today after a correspondence in which I submitted a crash/log or report to Mathworks:

"Hello Zachary,

After investigating this issue further, it seems that the crash is related to the MTEX Toolbox based on the stack trace "“/Users/Z/Documents/MATLAB/mtex-5.7.0/extern/nfft/nfsftmex.mexmaci64+00295854 nfsft_trafo+00002574".

Since the crash is related to the third-party toolbox which is not written by us, it is out of our technical support scope. Please work with Lisa and narrow down the root cause.

I will close the case for now. If the root cause is in MATLAB, please reply to this email. I will be happy to reopen the case and help you further.

Sincerely,
Bojil Ivanov
MathWorks Technical Support Department"


So, it seems that indeed the trouble is with MTEX nfft on my Mac and Windows installs. However, I spent almost the entire day yesterday trying to manually compile NFFT and openmp on my Mac, to no avail. However, I note that although my Mac install reports an issue at startup regarding running NFFT with openmp (see below), my Windows-based install does not repot that error, yet Matlab also crashes on that Windows computer.

Error message at startup...

initialize MTEX 5.7.0  ....
********************************************************************************
 
  MTEX: Error running NFFT with openmp - I'm going to use it without openmp!

********************************************************************************
 done!


In the meantime, my work-around is to use an older syntax I mentioned above and relist here:
`r = plotS2Grid('resolution',1*degree,'antipodal');`
`kde = calcDensity([g.longAxis],r,'halfwidth',10*degree);`
`[~,I]=max(kde)`
`V = r(I)`

I do not quite understand why one procedure/syntax works and the other doesn't, although I do recognize that the newer approach (which *only sometimes* crashes) produces a different class of variable (an S2FunHarmonic vs. a list/vector of KDE values) prior to the call to max().

I hope this is something that can be fixed in future releases someday.

Thanks for listening,
Zach

**Comment by zmichels** [comment]
Hi Rüdiger,

Thanks. Good to know.

Yes they do also use a mac.

I am on R2020a and MTEX 5.7.0
And it works fine with many sets of vectors and only crashes with some.

Zach

**Reply by kilir** [reply]
Do you use the pre-compiled mex binaries of nfft or did you build those yourself?

In the former case, maybe someone who can reproduce the issue, could try to build the nfft for their very machine (instructions: https://mtex-toolbox.github.io/installation#compiling-the-nfft) and see if this helps.

Cheers,
Rüdiger

**Comment by zmichels** [comment]
I use pre-compiled binaries, and MTEX says everything installs fine. However, when I startup Matlab (and MTEX), I get the following message:
"MTEX: Error running NFFT with openmp - I'm going to use it without openmp!”

I have a couple times tried/installed openmp, and it seems everything completes successfully, but I still get that message at startup.

I will try the NFFT install you suggest.

Thanks,
Zach

**Comment by zmichels** [comment]
Hi Ralf,

plot(S2Crash)

works fine and plots.

I then tried:
S2Crash.bandwidth=32

and, after that

max(S2Crash)

crashes again.

Zach

**Comment by zmichels** [comment]
Also... I just tried this on a Windows computer install in my SEM lab running r2021a and MTEX 5.7.0
This computer does not mention any errors about NFFT or openmp.
It also crashed Matlab on that computer.

**Reply by zmichels** [reply]
![image](https://user-images.githubusercontent.com/8769619/126357785-10c1a624-6985-422c-a098-7d39b09c0ac3.jpeg)

**Reply by zmichels** [reply]
HOWEVER!

On the Windows computer... if I do the:
S2Crash.bandwidth=32

then the max() function works and returns:

>> max(S2Crash)

ans =

    7.2526

**Reply by zmichels** [reply]
For whatever it is worth or in case it helps, I just installed Matlab (r2021a) and MTEX 5.7.0 on a second Windows computer (identical install procedure). Unlike the other Windows computer, Matlab does not crash on this computer. However, I do not have regular access to this computer, and it is unclear to me how the two are different, such that I am unsure how to update/fix the one that does crash.

**Reply by ralfHielscher** [reply]
Hi Zach,

I'n very sorry for all the incontinence this causes for your. Your crash is extremely strange as ```plot(S2Crash)```  calls exactly the same functions in the nfft library as ```max(S2Crash)```.

One clarification about openMP. MTEX comes with NFFT binaries with and without openMP support. At startup MTEX checks whether the binaries with openMP support are working. If not you get the warning an MTEX falls back to the binaries without openMP support. This does not mean any restriction in functionality.

When the binaries without openMP support crash as well (as in your case in some specific cases) you can try to recompile them on your computer. This might be more easy than compiling the binaries with openMP support and is explain here https://github.com/mtex-toolbox/mtex/tree/develop/extern/nfft

I hope we can find a solution, but I really do not know where to search.

All the best,

Ralf.

**Comment by zmichels** [comment]
Thanks, Ralf.

I agree that it is odd as you mention, and I will probably try to recompile the NFFT binaries at some point to see if that helps. 

Thank you also for clarification regarding openmp… I had suspected that was the case, but it is good to know for certain.

It is not much inconvenience for me to use the older syntax. My main concern regards sharing code with colleagues (that is how I discovered this issue). For now, I will continue to use the other syntax for snippets I share, to help ensure reproducibility across systems.

Cheers,
Zach

**Comment by zmichels** [comment]
Another odd/confusing data-point to share.

On the Windows machine that crashes... if I first run:
`plot(S2Crash)`
(plots successfully)

...and *then* try:
`max(S2Crash)`

... Matlab does *not* crash.

However, on my Mac, it makes no difference, and Matlab crashes after max(), even though it plots successfully. I have not yet had the time to try compiling the NFFT binaries. I'll report back if/when I get to that.

Also... today on the Windows computer that crashes, I tried 
`S2Crash.bandwidth = 32`
`max(S2Crash)`
... which yesterday worked without crashing. However today it causes a crash every time. I have made no changes to the install on that computer. So weird.

**Comment by lisakemp** [comment]
Zachary –

Thanks for sharing so much detail and for working through this. I’m sure it’s annoying.  You’re being very thorough and thoughtful.

You and Rudiger are much closer to the nitty gritty and way more knowledgeable about how MTEX is built than I. But it does sound like you’re on the right track, looking at versions of the toolbox, libraries, etc. And now how different PCs are set up differently, making life more complicated but still relevant.

Also consider if there are libraries in MTEX that could be clashing with other versions in MATLAB.

I’ll continue to follow the progress.

Lisa
